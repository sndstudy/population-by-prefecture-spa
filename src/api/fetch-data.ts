import axios from 'axios';

const RESAS_API_ENDPOINT = 'https://opendata.resas-portal.go.jp/';
const API_KEY = process.env['REACT_APP_RESAS_API_KEY'] || '';

const instance = axios.create({
  baseURL: RESAS_API_ENDPOINT,
  timeout: 1000,
  headers: { 'X-API-KEY': API_KEY },
});

export type PrefecturesResponse = {
  message: string;
  result: Prefectures[];
};

export type Prefectures = {
  prefCode: string;
  prefName: string;
};

export type PopulationResponse = {
  message: string;
  result: Population;
};

export type Population = {
  boundaryYear: number;
  data: {
    label: string;
    data: {
      year: number;
      value: number;
      rate?: number;
    }[];
  }[];
};

export const fetchPrefectures = async () => {
  const { data } = await instance.get<
    PrefecturesResponse | { statusCode: string }
  >('api/v1/prefectures');

  // エラーの場合、ステータスコードがレスポンスボディに設定されているため「statusCode」の有無でエラーを判定する
  // https://opendata.resas-portal.go.jp/docs/api/v1/detail/index.html
  if ('statusCode' in data) {
    return Promise.reject();
  }

  return data;
};

export const fetchPopulationData = async ({
  queryKey,
}: {
  queryKey: string[];
}) => {
  const [, prefCode] = queryKey;

  const { data } = await instance.get<
    PopulationResponse | { statusCode: string }
  >('api/v1/population/composition/perYear', {
    params: {
      prefCode,
    },
  });

  // エラーの場合、ステータスコードがレスポンスボディに設定されているため「statusCode」の有無でエラーを判定する
  // https://opendata.resas-portal.go.jp/docs/api/v1/detail/index.html
  if ('statusCode' in data) {
    return Promise.reject();
  }

  return data;
};
