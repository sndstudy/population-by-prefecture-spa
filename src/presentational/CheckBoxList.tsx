import { ChangeEventHandler } from 'react';
import { Prefectures } from '../api/fetch-data';
import CheckBox from './CheckBox';

type Props = {
  data: Prefectures[];
  selectedPrefectures: string[];
  isDisabled: boolean;
  onChange: ChangeEventHandler<HTMLInputElement>;
};

const CheckBoxList = (props: Props) => {
  return (
    <>
      {props.data.map((prefecture) => {
        return (
          <CheckBox
            key={prefecture.prefCode}
            value={prefecture.prefCode}
            label={prefecture.prefName}
            checked={props.selectedPrefectures.some(
              (selectedPrefecture) =>
                selectedPrefecture === String(prefecture.prefCode),
            )}
            isDisabled={props.isDisabled}
            onChange={props.onChange}
          />
        );
      })}
    </>
  );
};

export default CheckBoxList;
