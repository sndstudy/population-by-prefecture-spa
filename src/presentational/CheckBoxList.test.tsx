import { render, screen } from '@testing-library/react';
import { Prefectures } from '../api/fetch-data';
import CheckBoxList from './CheckBoxList';

describe('CheckBoxListのテスト', () => {
  it('データが0件の場合', () => {
    const data: Prefectures[] = [];
    const selectedPrefectures: string[] = [];
    const { asFragment } = render(
      <CheckBoxList
        data={data}
        selectedPrefectures={selectedPrefectures}
        isDisabled={false}
        onChange={jest.fn}
      />,
    );
    const checkBoxList = screen.queryAllByRole('checkbox');
    expect(checkBoxList.length).toBe(0);
    expect(asFragment()).toMatchSnapshot();
  });

  it('データがある場合。(データ全てが選択されている状態)', () => {
    const data: Prefectures[] = [
      { prefCode: '1', prefName: '北海道' },
      { prefCode: '2', prefName: '青森県' },
      { prefCode: '3', prefName: '岩手県' },
    ];
    const selectedPrefectures: string[] = ['1', '2', '3'];
    const { asFragment } = render(
      <CheckBoxList
        data={data}
        selectedPrefectures={selectedPrefectures}
        isDisabled={false}
        onChange={jest.fn}
      />,
    );
    const checkBoxList = screen.queryAllByRole('checkbox');
    expect(checkBoxList.length).toBe(3);
    checkBoxList.forEach((v) => {
      expect(v).toBeChecked();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('データがある場合。(データ全てが選択されていない状態)', () => {
    const data: Prefectures[] = [
      { prefCode: '1', prefName: '北海道' },
      { prefCode: '2', prefName: '青森県' },
      { prefCode: '3', prefName: '岩手県' },
    ];
    const selectedPrefectures: string[] = [];
    const { asFragment } = render(
      <CheckBoxList
        data={data}
        selectedPrefectures={selectedPrefectures}
        isDisabled={false}
        onChange={jest.fn}
      />,
    );
    const checkBoxList = screen.queryAllByRole('checkbox');
    expect(checkBoxList.length).toBe(3);
    checkBoxList.forEach((v) => {
      expect(v).not.toBeChecked();
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
