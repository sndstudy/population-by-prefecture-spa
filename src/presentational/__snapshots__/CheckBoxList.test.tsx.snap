// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBoxListのテスト データが0件の場合 1`] = `<DocumentFragment />`;

exports[`CheckBoxListのテスト データがある場合。(データ全てが選択されていない状態) 1`] = `
<DocumentFragment>
  <div>
    <label>
      <input
        type="checkbox"
        value="1"
      />
      北海道
    </label>
  </div>
  <div>
    <label>
      <input
        type="checkbox"
        value="2"
      />
      青森県
    </label>
  </div>
  <div>
    <label>
      <input
        type="checkbox"
        value="3"
      />
      岩手県
    </label>
  </div>
</DocumentFragment>
`;

exports[`CheckBoxListのテスト データがある場合。(データ全てが選択されている状態) 1`] = `
<DocumentFragment>
  <div>
    <label>
      <input
        checked=""
        type="checkbox"
        value="1"
      />
      北海道
    </label>
  </div>
  <div>
    <label>
      <input
        checked=""
        type="checkbox"
        value="2"
      />
      青森県
    </label>
  </div>
  <div>
    <label>
      <input
        checked=""
        type="checkbox"
        value="3"
      />
      岩手県
    </label>
  </div>
</DocumentFragment>
`;
